casanovo/denovo/model.py

```diff
@@ -989,6 +989,190 @@ def configure_optimizers(
         return [optimizer], {"scheduler": lr_scheduler, "interval": "step"}
 
 
+class DBSpec2Pep(Spec2Pep):
+    """
+    Inherits Spec2Pep
+
+    Hijacks teacher-forcing implemented in Spec2Pep and
+    uses it to predict scores between a spectra and associated peptide.
+    Input format is .mgf, with comma-separated targets
+    and decoys in the SEQ field. Decoys should have a prefix of "decoy_".
+    """
+
+    num_pairs = None  # Modified to be predict_batch_size from config
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def predict_step(self, batch, *args):
+        batch_res = []
+        for (
+            indexes,
+            t_or_d,
+            peptides,
+            precursors,
+            encoded_ms,
+        ) in self.smart_batch_gen(batch):
+            pred, truth = self.decoder(peptides, precursors, *encoded_ms)
+            sm = torch.nn.Softmax(dim=2)
+            pred = sm(pred)
+            score_result, per_aa_score = _calc_match_score(
+                pred, truth
+            )  # Calculate the score between spectra + peptide list
+            batch_res.append(
+                (
+                    indexes,
+                    t_or_d,
+                    peptides,
+                    score_result.cpu().detach().numpy(),
+                    per_aa_score.cpu().detach().numpy(),
+                    precursors.cpu().detach().numpy(),
+                )
+            )
+        return batch_res
+
+    def smart_batch_gen(self, batch):
+        all_psm = []
+        enc = self.encoder(batch[0])
+        precursors = batch[1]
+        indexes = batch[3]
+        enc = list(zip(*enc))
+        for idx, _ in enumerate(batch[0]):
+            spec_peptides = batch[2][idx].split(",")
+            # Check for decoy prefixes and create a bit-vector indicating targets (1) or decoys (0)
+            decoy_prefix = "decoy_"  # Decoy prefix
+            t_or_ds = [
+                0 if p.startswith(decoy_prefix) else 1 for p in spec_peptides
+            ]
+            # Remove decoy prefix
+            spec_peptides = [
+                s[len(decoy_prefix) :] if s.startswith(decoy_prefix) else s
+                for s in spec_peptides
+            ]
+            spec_precursors = [precursors[idx]] * len(spec_peptides)
+            spec_enc = [enc[idx]] * len(spec_peptides)
+            spec_idx = [indexes[idx]] * len(spec_peptides)
+            all_psm.extend(
+                list(
+                    zip(
+                        spec_enc,
+                        spec_precursors,
+                        spec_peptides,
+                        spec_idx,
+                        t_or_ds,
+                    )
+                )
+            )
+        # Continually grab num_pairs items from all_psm until list is exhausted
+        while len(all_psm) > 0:
```

Comment:
Ideally, this logic should be moved to the dataloader, and predict step should just receive batches of eval_batch_size psms.